src/functional/program_processors/codegens/gtfn/gtfn_ir_to_gtfn_im_ir.py

```diff
@@ -0,0 +1,274 @@
+                if node.id == "nbh_iter":
+                    return self.cur_idx
+                return self.generic_visit(node)
+
+            def __init__(self, cur_idx):
+                self.cur_idx = cur_idx
+
+        for i in range(max_neighbors):
+            new_expr = PlugInCurrentIdx(cur_idx=gtfn_ir.OffsetLiteral(value=i)).visit(new_fun)
+            rhs = self.visit(new_expr, **kwargs)
+            self.imp_list_ir.append(AssignStmt(lhs=gtfn_ir_common.SymRef(id=red_idx), rhs=rhs))
+
+        return gtfn_ir_common.SymRef(id=red_idx)
+
+    def visit_FunCall(self, node: gtfn_ir.FunCall, **kwargs):
+        if isinstance(node.fun, gtfn_ir.Lambda) and any(
+            isinstance(
+                arg,
+                gtfn_ir.Lambda,
+            )
+            for arg in node.args
+        ):
+            # do not try to lower lambdas that take lambdas as arugment to something more readable
+            red_idx = self.uids.sequential_id(prefix="red")
+            self.imp_list_ir.append(InitStmt(lhs=gtfn_ir_common.Sym(id=f"{red_idx}"), rhs=node))
+            return gtfn_ir_common.SymRef(id=f"{red_idx}")
+        if isinstance(node.fun, gtfn_ir.Lambda):
+            lam_idx = self.uids.sequential_id(prefix="lam")
+            params = [self.visit(param, **kwargs) for param in node.fun.params]
+            args = [self.visit(arg, **kwargs) for arg in node.args]
+            for param, arg in zip(params, args):
+                if param.id in self.sym_table:
+                    self.localized_symbols.append(param.id)
+                self.imp_list_ir.append(
+                    InitStmt(
+                        lhs=gtfn_ir_common.Sym(id=f"{param.id}_local")
+                        if param.id in self.sym_table
+                        else gtfn_ir_common.Sym(id=f"{param.id}"),
+                        rhs=arg,
+                    )
+                )
+            expr = self.visit(node.fun.expr, **kwargs)
+            self.imp_list_ir.append(InitStmt(lhs=gtfn_ir_common.Sym(id=f"{lam_idx}"), rhs=expr))
+            return gtfn_ir_common.SymRef(id=f"{lam_idx}")
+        if _is_reduce(node):
+            return self.handle_Reduction(node, **kwargs)
+        if (
+            isinstance(node.fun, gtfn_ir_common.SymRef) and node.fun.id == "make_tuple"
+        ):  # TODO: bad hardcoded string
+            tupl_idx = self.uids.sequential_id(prefix="tupl")
+            for i, arg in enumerate(node.args):
+                expr = self.visit(arg, **kwargs)
+                self.imp_list_ir.append(
+                    InitStmt(lhs=gtfn_ir_common.Sym(id=f"{tupl_idx}_{i}"), rhs=expr)
+                )
+            tup_args = [gtfn_ir_common.SymRef(id=f"{tupl_idx}_{i}") for i in range(len(node.args))]
+            tuple_fun = gtfn_ir.FunCall(fun=gtfn_ir_common.SymRef(id="make_tuple"), args=tup_args)
+            self.imp_list_ir.append(
+                InitStmt(lhs=gtfn_ir_common.Sym(id=f"{tupl_idx}"), rhs=tuple_fun)
+            )
```

Comment:
it seems like this could work in general for a any function, right? i.e. 
```
fun(a, b)
```
to 
```
var_a = a
var_b = b
var res = fun(var_a, var_b)
```
If we extract this into a function it might be possible to reuse it in other places.